### DISCUSSION TITLE
Wireshark (Pcap) Capture?

### DISCUSSION THREAD

**Discussion by xandfury** [aiortc/aiortc#678] [Q&A] [answered]
Hi. Thanks for `aiortc`. I am a newbie to `webrtc`. So apologies in advance! 😅 I am trying to understand the webrtc protocol. For the same, I ran the `datachannel-cli` example and I see the flow of messages (with verbose on). But for some reason, I can't find `SCTP` transmission in wireshark. Any ideas why that is? I am looking with sctp filter

```
DEBUG:aioice.ice:Connection(0) protocol(0) < ('192.168.1.143', 58744) Message(message_method=Method.BINDING, message_class=Class.REQUEST, transaction_id=b'\xe1\x8e\xce\xb5<)\x9b\xab\x9cd\xfc\xb1')
DEBUG:aioice.ice:Connection(0) protocol(0) > ('192.168.1.143', 58744) Message(message_method=Method.BINDING, message_class=Class.RESPONSE, transaction_id=b'\xe1\x8e\xce\xb5<)\x9b\xab\x9cd\xfc\xb1')
DEBUG:aiortc.rtcsctptransport:RTCSctpTransport(server) < DataChunk(flags=3, tsn=809524935, stream_id=1, stream_seq=44)
DEBUG:aiortc.rtcsctptransport:RTCSctpTransport(server) > SackChunk(flags=0, advertised_rwnd=1048576, cumulative_tsn=809524935, gaps=[])
DEBUG:aiortc.rtcsctptransport:RTCSctpTransport(server) > DataChunk(flags=3, tsn=3702826693, stream_id=1, stream_seq=44)
DEBUG:aiortc.rtcsctptransport:RTCSctpTransport(server) - T3 start
DEBUG:aiortc.rtcsctptransport:RTCSctpTransport(server) < SackChunk(flags=0, advertised_rwnd=1048576, cumulative_tsn=3702826693, gaps=[])
DEBUG:aiortc.rtcsctptransport:RTCSctpTransport(server) - T3 cancel
channel(chat) < ping 43062094
channel(chat) > pong 43062094
```

**Comment by xandfury** [comment, marked as answer]
oh wait. I just read that sctp is encapsulated within udp for `webrtc`. Nvm. Thanks!

**Reply by xandfury** [reply]
I can see `UDP` and `DTLS` in wireshark. But need a to find a way to decrypt the payload..

**Reply by jlaine** [reply]
Right, the traffic is indeed encrypted!

It *might* be possible to modify `aiortc` to export the secrets for Wireshark to use, but I'm not sure whether this would even work for DTLS. FYI, most TLS stacks provide support for exporting secrets, triggered by an `SSLKEYLOGFILE` environment variable

**Comment by xandfury** [comment]
Thanks for the tip @jlaine ! [I looked at the docs](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_keylog_cb_func.html). This may work 🙂 

I added the following [here](https://github.com/aiortc/aiortc/blob/main/src/aiortc/rtcdtlstransport.py#L201-L222): 
```python
        # Configure elliptic curve for ECDHE in server mode for OpenSSL < 1.1.0
        if lib.OpenSSL_version_num() < 0x10100000:  # pragma: no cover
            lib.SSL_CTX_set_ecdh_auto(ctx, 1)

        if keylog:               # this is just a debug flag which is toggleable
            lib.SSL_CTX_set_keylog_callback(ctx, keylog_callback)
        return ctx
```

with the following callback function:
```python
@ffi.callback("void (*)(const SSL *, const char *)")
def keylog_callback(x, y):
    if y:
        print(ffi.string(y).decode("utf-8"))
```

gives something like:
```
'CLIENT_RANDOM b13b4c3377230dd4e3c651354d7fc2df594d85cc69862ec9c29d4e6d0d36be64 d74c83e74e8a5579926523f20187cc4e4d375ebf09c4d35f8f1f403dd9cd2ca89e40294d589b8d7650550a29ced506d8'
```

**Reply by xandfury** [reply]
I may have celebrated too early. This does not work with DTLS.

**Reply by jlaine** [reply]
I'm not familiar with what OpenSSL provides for logging DTLS keys, maybe "exporter secret" is another keyword you might want to look up. Keep going though, I'm sure you'll crack and I would be happy to add this feature once you have it figured out.
